### PR TITLE
Use Crate's Postgres support, introduce config file

### PR DESCRIPTION
**Pull request by juliusv**
Fixes https://github.com/crate/crate_adapter/issues/14
Fixes https://github.com/crate/crate_adapter/issues/13
Fixes https://github.com/crate/crate_adapter/issues/12

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -6,7 +6,7 @@ This is an adapter that accepts Prometheus remote read/write requests,
 and sends them on to CrateDB. This allows using CrateDB as long term storage
 for Prometheus.
 
-Requires CrateDB 2.2.0 or greater.
+Requires CrateDB 3.1.0 or greater.
```
Comment (juliusv) [inline README.rst]: Btw.,  CrateDB 3.1.0 is not out yet, but my nightly build that has the required features says `3.1.0-SNAPSHOT` on startup, so I assume it'll be 3.1.0 once it's released.

Comment (chaudum) [reply README.rst]: Yes, the BEGIN/COMMIT feature + the fix for the TIMESTAMPTZ type will be available in CrateDB 3.1.0.

**Comment by chaudum** [comment]
Is it really best practice to add dependencies in Github? I'm not a Golang expert, but that does not feel right. Dependencies can be resolved when the project is built, right?

File: crate.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+		case *crateReadRequest:
+			return c.read(ctx, r)
+		default:
+			panic("unknown request type")
+		}
+	}
+}
+
+func (c crateEndpoint) write(ctx context.Context, r *crateWriteRequest) error {
+	_, err := c.pool.PrepareEx(ctx, "write_statement", `INSERT INTO metrics ("labels", "labels_hash", "timestamp", "value", "valueRaw") VALUES ($1, $2, $3, $4, $5)`, nil)
```
Comment (chaudum) [inline crate.go]: I think we should make the prepared statement name unique, e.g. suffix with timestamp.

Comment (chaudum) [reply crate.go]: Question: In Go, how performant is it to allocate strings?
In other languages I would use a constant for the statement, as this part of the code is executed very often.

**Comment by juliusv** [comment]
@chaudum Vendoring (making a copy) of all dependencies is pretty standard in Go for non-experimental projects (in repos that represent final binaries, not in library repos). It's the only way to ensure that your code can still be built - and built reasonably reproducibly - in X amount of time.

File: crate.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+	poolConf pgx.ConnPoolConfig
+}
+
+func newCrateEndpoint(ep *endpointConfig) *crateEndpoint {
+	connConf := pgx.ConnConfig{
+		Host:     ep.Host,
+		Port:     ep.Port,
+		User:     ep.User,
+		Password: ep.Password,
+		Database: ep.Database,
```
Comment (chaudum) [inline crate.go]: `Database: ep.Schema,`

CrateDB does not know the concept of a database in terms of other RDBMSs.
We should always say `schema`, if the equivalent of database in Postgres is meant.

**Comment by juliusv** [comment]
Some more background on this:

https://peter.bourgon.org/go-best-practices-2016/#dependency-management
https://peter.bourgon.org/go-in-production/#dependency-management

File: crate.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+	}
+
+	batch := c.pool.BeginBatch()
+	for _, a := range r.rows {
+		batch.Queue(
+			"write_statement",
+			[]interface{}{
+				a.labels,
+				a.labelsHash,
+				// TODO: Find non-string way of encoding timestamps.
```
Comment (chaudum) [inline crate.go]: We will add a QA test for all types https://github.com/crate/crate-qa/blob/master/tests/client_tests/go/bulk_operations.go

**Comment by juliusv** [comment]
(note that the details around the mechanics of vendoring in the second link are pretty outdated)

File: server.go [hunk 5]
```diff
@@ -290,37 +266,28 @@ func (ca *crateAdapter) handleRead(w http.ResponseWriter, r *http.Request) {
...
-			args = append(args, int64(math.Float64bits(s.Value)))
-			args = append(args, s.TimestampMs)
-
-			request.BulkArgs = append(request.BulkArgs, args)
+			request.rows = append(request.rows, &crateRow{
+				labels:     metric,
+				labelsHash: fp,
+				timestamp:  time.Unix(0, s.Timestamp*1e6).UTC(),
+				value:      s.Value,
+				// Crate.io can't handle full NaN values as required by Prometheus 2.0,
```
Comment (chaudum) [inline server.go]: We would like to fix this. What exactly does CrateDB not support? What does "full NaN" mean?

File: server.go [hunk 6]
```diff
@@ -365,50 +332,64 @@ func (ca *crateAdapter) handleWrite(w http.ResponseWriter, r *http.Request) {
...
 }
 
-func encodeCrateRequest(_ context.Context, r *http.Request, request interface{}) error {
-	jsonRequest, err := json.Marshal(request)
+type endpointConfig struct {
+	Host             string `yaml:"host"`
+	Port             uint16 `yaml:"port"`
+	User             string `yaml:"user"`
+	Password         string `yaml:"password"`
+	Database         string `yaml:"database"`
```
Comment (chaudum) [inline server.go]: Please name the setting "schema". See first comment

File: crate.go [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+				a.timestamp.Format("2006-01-02 15:04:05.000-07"),
```
[the remark this answers is not in the document]
Comment (juliusv) [reply crate.go]: Great, will test and update here if it works.

File: crate.go (hunk 2 above)
Comment (juliusv) [reply crate.go]: > I think we should make the prepared statement name unique, e.g. suffix with timestamp.

Reading https://godoc.org/github.com/jackc/pgx#ConnPool.PrepareEx and skimming the underlying implementation, it at least seems like it should be ok to always use the same name for the same statement string:

"PrepareEx is idempotent; i.e. it is safe to call PrepareEx multiple times with the same name and sql arguments. This allows a code path to PrepareEx and Query/Exec/Prepare without concern for if the statement has already been prepared."

> Question: In Go, how performant is it to allocate strings?
In other languages I would use a constant for the statement, as this part of the code is executed very often.

I'd be surprised if you could measure any difference (haven't tried yet), especially since this only happens once per remote-write-request, not per sample. But I think it's a good idea to put it into a `const` anyway.

File: crate.go (hunk 3 above)
Comment (juliusv) [reply crate.go]: Good to know, will call it schema on our end (can't change the `pgx.ConnConfig` struct of course).

File: server.go (hunk 5 above)
Comment (juliusv) [reply server.go]: This is a comment that already existed in `master`, but I'm sure it's about the difference between "normal" NaN values and signalling NaN values, see https://github.com/prometheus/prometheus/blob/master/pkg/value/value.go and https://en.wikipedia.org/wiki/NaN#Signaling_NaN.

Prometheus uses one type of signalling NaN to explicitly mark stale samples. For this to work, the underlying database has to support storing the exact bit pattern that is sent to it. Not sure if CrateDB can do that via Postgres now?

File: crate.go (hunk 2 above)
Comment (chaudum) [reply crate.go]: Ok, cool. Sounds good to me.

**Comment by chaudum** [comment]
@juliusv Regarding vendoring. Happy to learn something new :) However, makes perfectly sense, as there aren't any mirrored Go package repositories that would guarantee long-term availability.

File: crate.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (juliusv) [reply crate.go]: Hmm, I tried two variants here, both based against the latest night build based on 33f0c20, which includes commit 90e4343.

This one gives the error: `cannot convert {2018-07-03 13:54:21.99 +0000 UTC 2 none} to Timestamptz`:

```golang
pgtype.Timestamptz{Time: a.timestamp, Status: pgtype.Present}
```

This pointer variant seems to succeed, but all timestamps in the DB end up being `946684800000 (2000-01-01T00:00:00.000Z)`:

```golang
&pgtype.Timestamptz{Time: a.timestamp, Status: pgtype.Present}
```

Do you perhaps have a working example in Go?

[the remark this answers is not in the document]
Comment (juliusv) [reply crate.go]: If I just try `a.timestamp` btw., which is just a `time.Time`, I also get only `2000-01-01T00:00:00.000` inserted.

[the remark this answers is not in the document]
Comment (chaudum) [reply crate.go]: Think we'll need to investigate further. We have a test for inserting only, but we don't check the actual value that we store. We should've done that really.
Think we can keep the "workaround" (as it's universally working across versions) until we've a fix for the correct timestamp insert.

[the remark this answers is not in the document]
Comment (juliusv) [reply crate.go]: Ok cool, I pushed the other fixups now.

**Comment by chaudum** [comment]
@joemoe this PR looks good to me. should we discuss release strategy before we merge into master?

**Comment by chaudum** [comment]
@juliusv For the future we may want to look at https://github.com/golang/dep for dependency management/vendoring. Do you know this project?

**Comment by chaudum** [comment]
@juliusv Could you squash the last 3 commits + the review fixups into a single commit, so there are only 2 commits in total. Then I'm good to merge it into master.
We'll do a new release then.

**Comment by juliusv** [comment]
> For the future we may want to look at https://github.com/golang/dep for dependency management/vendoring. Do you know this project?

dep was built by a group of experienced Go people and once slated to become the official vendoring solution for Go. But then some people from the original Go team found that they were not happy with some aspects of dep and created yet another tool, vgo (https://github.com/golang/vgo, https://golang.org/issue/24301), that the Go team now decided should become the standard Go vendoring tool. There was a big community uprising against that, as there were people unhappy with the way vgo resolves versions of dependencies, and it all turned into a religious war... so the final outcome isn't really clear yet, except that vgo probably has a higher probability of actually turning into the "official" vendoring tool that will be built into Go itself. Personally I haven't used vgo yet, but have used dep and think it's a very solid solution. When using dep you should be aware though that it takes much longer to run than say "govendor", since it does more sophisticated dependency resolution and then downloads the necessary package versions on every vendoring update, rather than just copying what you have in your GOPATH somewhere.

In summary: it shouldn't matter too much, you can switch easily between vendoring tools, the current govendor works fine enough too, but dep might be a more robust solution, vgo might become the standard though, but it's early :)

**Comment by juliusv** [comment]
@chaudum Squashed the commits as requested (now there's a vendoring commit and one with all the other changes).

**Comment by juliusv** [comment]
I added tests for the configuration loading (which actually found some issues around YAML parsing and defaults). I squashed it all again immediately so it'd be ready to merge, I hope that's ok.

**Comment by chaudum** [comment]
@juliusv Thanks a lot!
